src/drivers/amd/renoir/ipc.c

```diff
@@ -163,8 +182,20 @@ int ipc_platform_send_msg(struct ipc_msg *msg)
 		sw_intr_stat = (acp_dsp_sw_intr_stat_t)
 				io_reg_read(PU_REGISTER_BASE + ACP_DSP_SW_INTR_STAT);
 		status =  sw_intr_stat.bits.dsp0_to_host_intr_stat;
-		return ret;
+		ret = -EBUSY;
+		goto out;
+	}
+	lock = io_reg_read(PU_REGISTER_BASE + ACP_AXI2DAGB_SEM_0);
+	while (lock) {
+		lock = io_reg_read(PU_REGISTER_BASE + ACP_AXI2DAGB_SEM_0);
+		if (!delay_cnt) {
+			ret = -EBUSY;
+			break;
+		}
+		delay_cnt--;
```

Comment:
The code looks a bit weird, you have two reads from the same register within two lines.

Could a do.. while loop simplify?

Reply:
This is logic is implemented based on our hardware behavior.
This register has special behavior for reads. If this bit is currently 0, a read will set this bit to 1 by hardware and 0 is returned for the read which will acquire the semaphore. If this bit is currently 1, a read will return 1 and the bit remains unchanged.  Writes have normal behavior.

Reply:
> The code looks a bit weird, you have two reads from the same register within two lines.

The first read is to check whether the lock is acquired or not, if the lock is already acquired by other module, then the second read within the loop is till the lock is released by other module or timeout occurs.